File: apps/menu-bar/Sources/YTMMenuBarConnector/ConnectorApp.swift
import AppKit
import Foundation

final class ConnectorApp {
  private static let playbackStateRetryDelaySeconds: TimeInterval = 2
  private static let playbackStateStaleTimeoutSeconds: TimeInterval = 8
  private static let staleProgressToleranceSeconds: Double = 0.25

  private let connection: ConnectorConnection
  private let menu: MenuBarController
  private let logger: NativeAppLogger
  private var nextRequestNumber = 0
  private var isReady = false
  private var playbackStateRetry: DispatchWorkItem?
  private var playbackStateStaleTimeout: DispatchWorkItem?
  private var lastAcceptedPlaybackState: PlaybackState?
  private var isPlaybackStateStale = false
  var onRequestUninstall: (() -> Void)?

  init(
    connection: ConnectorConnection,
    menu: MenuBarController,
    logger: NativeAppLogger = NativeAppLogger()
  ) {
    self.connection = connection
    self.menu = menu
    self.logger = logger
  }

  func start() {
    menu.onShuffle = { [weak self] in self?.sendAction("shuffle") }
    menu.onPrevious = { [weak self] in self?.sendAction("previous") }
    menu.onTogglePlay = { [weak self] in self?.sendAction("togglePlay") }
    menu.onNext = { [weak self] in self?.sendAction("next") }
    menu.onRepeat = { [weak self] in self?.sendAction("repeat") }
    menu.onSeek = { [weak self] time in self?.sendSeek(time) }
    menu.onFocusYouTubeMusic = { [weak self] in self?.sendFocusYouTubeMusic() }
    logger.log("connector app starting logPath=\(logger.path)")

    connection.start(
      onMessage: { [weak self] message in
        self?.handle(message)
      },
      onDisconnect: { [weak self] in
        self?.logger.log("connector disconnected")
        self?.isReady = false
        self?.isPlaybackStateStale = false
        self?.lastAcceptedPlaybackState = nil
        self?.clearPlaybackStateRetry()
        self?.clearPlaybackStateStaleTimeout()
        self?.menu.updateConnectionStatus("Disconnected")
      }
    )

    connection.send(ConnectorProtocol.hello(requestId: nextRequestId("hello")))
  }

  private func handle(_ json: [String: Any]) {
    guard let message = HostMessage(json: json) else {
      let type = json["type"] as? String ?? "unknown"
      logger.log("received message could not be decoded type=\(type)")
      return
    }

    logger.log(
      "received message handling type=\(message.type) requestId=\(message.requestId ?? "none")"
    )

    switch message.type {
    case "connector.ready":
      isReady = true
      clearPlaybackStateStaleTimeout()
      menu.updateConnectionStatus("Connected")
      logger.log("connector ready; subscribing to playback state")
      connection.send(
        ConnectorProtocol.subscribePlayback(
          requestId: nextRequestId("subscribe")
        )
      )
      requestPlaybackState()
    case "playback.state":
      if let state = message.state {
        handlePlaybackState(state)
      } else {
        logger.log("playback state message missing state payload")
      }
    case "connector.error":
      let label = userFacingStatus(code: message.code, message: message.message)
      logger.log("connector error \(label)")
      if message.code == "connector_not_registered" {
        restartHandshake(reason: label)
        return
      }
      if isPlaybackStateRequestError(message) {
        logger.log("Waiting for YouTube Music")
        menu.updateConnectionStatus(label)
        schedulePlaybackStateRetry(reason: label)
        return
      }
      if isConnectorAvailabilityError(message.code) {
        isReady = false
        isPlaybackStateStale = false
        lastAcceptedPlaybackState = nil
        clearPlaybackStateRetry()
        clearPlaybackStateStaleTimeout()
      }
      menu.updateConnectionStatus(label)
    case ConnectorProtocol.uninstallRequestedType:
      logger.log("connector uninstall requested")
      onRequestUninstall?()
    default:
      logger.log("received message ignored type=\(message.type)")
      break
    }
  }

  private func handlePlaybackState(_ state: PlaybackState) {
    clearPlaybackStateRetry()
    clearPlaybackStateStaleTimeout()
    logger.log(playbackStateSummary(state))

    if shouldKeepStalePlaybackState(state) {
      logger.log("playback state still stale; ignoring non-advancing playing poll")
      menu.setStalePlaybackState()
      schedulePlaybackStateStaleTimeout()
      return
    }

    isPlaybackStateStale = false
    lastAcceptedPlaybackState = state
    menu.updatePlayback(state)

    if state.isPlaying {
      schedulePlaybackStateStaleTimeout()
    }
  }

  private func requestPlaybackState() {
    guard isReady else {
      logger.log("playback state refresh skipped because connector is not ready")
      return
    }
    clearPlaybackStateRetry()
    logger.log("requesting playback state")
    connection.send(
      ConnectorProtocol.playbackStateRequest(requestId: nextRequestId("state"))
    )
  }

  private func sendAction(_ action: String) {
    guard isReady else {
      logger.log("playback action skipped because connector is not ready")
      return
    }
    logger.log("sending playback action \(action)")
    connection.send(
      ConnectorProtocol.playbackAction(
        action,
        requestId: nextRequestId("action")
      )
    )
  }

  private func sendSeek(_ time: Double) {
    guard isReady else {
      logger.log("playback seek skipped because connector is not ready")
      return
    }
    logger.log("sending playback seek time=\(time)")
    connection.send(
      ConnectorProtocol.playbackSeek(
        time: time,
        requestId: nextRequestId("seek")
      )
    )
  }

  private func sendFocusYouTubeMusic() {
    guard isReady else {
      logger.log("focus YouTube Music skipped because connector is not ready")
      return
    }
    logger.log("sending focus YouTube Music")
    connection.send(
      ConnectorProtocol.focusYouTubeMusic(
        requestId: nextRequestId("focus")
      )
    )
  }

  private func nextRequestId(_ prefix: String) -> String {
    nextRequestNumber += 1
    return "\(prefix)-\(nextRequestNumber)"
  }

  private func playbackStateSummary(_ state: PlaybackState) -> String {
    [
      "playback state",
      "title=\(logValue(state.title))",
      "artist=\(logValue(state.artist))",
      "album=\(logValue(state.album))",
      "artwork=\(logValue(state.artworkUrl))",
      "nextTrack=\(logValue(state.nextTrack?.title))",
      "nextTrackArtwork=\(logValue(state.nextTrack?.artworkUrl))",
      "year=\(state.year.map(String.init) ?? "nil")",
      "playing=\(state.isPlaying)",
      "progress=\(Int(state.progress.rounded()))",
      "duration=\(Int(state.duration.rounded()))",
      "shuffle=\(state.isShuffling.map(String.init) ?? "nil")",
      "repeat=\(state.repeatMode ?? "nil")",
    ].joined(separator: " ")
  }

  private func isPlaybackStateRequestError(_ message: HostMessage) -> Bool {
    guard message.requestId?.hasPrefix("state-") == true else {
      return false
    }

    return isMissingYouTubeMusicTabError(message.message)
  }

  private func schedulePlaybackStateRetry(reason: String) {
    clearPlaybackStateRetry()
    let retry = DispatchWorkItem { [weak self] in
      self?.playbackStateRetry = nil
      self?.logger.log("playback state retry starting")
      self?.requestPlaybackState()
    }
    playbackStateRetry = retry
    logger.log(
      "playback state retry scheduled reason=\(reason) delay=\(Self.playbackStateRetryDelaySeconds)"
    )
    DispatchQueue.main.asyncAfter(
      deadline: .now() + Self.playbackStateRetryDelaySeconds,
      execute: retry
    )
  }

  private func schedulePlaybackStateStaleTimeout() {
    clearPlaybackStateStaleTimeout()
    let timeout = DispatchWorkItem { [weak self] in
      self?.playbackStateStaleTimeout = nil
      self?.markPlaybackStateStale()
    }
    playbackStateStaleTimeout = timeout
    logger.log(
      "playback state stale timeout scheduled delay=\(Self.playbackStateStaleTimeoutSeconds)"
    )
    DispatchQueue.main.asyncAfter(
      deadline: .now() + Self.playbackStateStaleTimeoutSeconds,
      execute: timeout
    )
  }

  private func markPlaybackStateStale() {
    isPlaybackStateStale = true
    logger.log("playback state stale; marking controls as reconnecting")
    menu.setStalePlaybackState()
    requestPlaybackState()
  }

  private func restartHandshake(reason: String) {
    isReady = false
    isPlaybackStateStale = false
    lastAcceptedPlaybackState = nil
    clearPlaybackStateRetry()
    clearPlaybackStateStaleTimeout()
    logger.log("connector handshake restarting reason=\(reason)")
    menu.updateConnectionStatus(reason)
    connection.send(ConnectorProtocol.hello(requestId: nextRequestId("hello")))
  }

  private func clearPlaybackStateRetry() {
    playbackStateRetry?.cancel()
    playbackStateRetry = nil
  }

  private func clearPlaybackStateStaleTimeout() {
    playbackStateStaleTimeout?.cancel()
    playbackStateStaleTimeout = nil
  }

  private func isConnectorAvailabilityError(_ code: String?) -> Bool {
    switch code {
    case "host_disabled", "connector_blocked", "unsupported_protocol":
      return true
    default:
      return false
    }
  }

  private func shouldKeepStalePlaybackState(_ state: PlaybackState) -> Bool {
    guard
      isPlaybackStateStale,
      state.isPlaying,
      state.duration > 0,
      let previous = lastAcceptedPlaybackState,
      samePlaybackItem(previous, state)
    else {
      return false
    }

    return abs(state.progress - previous.progress) <= Self.staleProgressToleranceSeconds
  }

  private func samePlaybackItem(_ lhs: PlaybackState, _ rhs: PlaybackState) -> Bool {
    normalizedPlaybackValue(lhs.title) == normalizedPlaybackValue(rhs.title)
      && normalizedPlaybackValue(lhs.artist) == normalizedPlaybackValue(rhs.artist)
      && normalizedPlaybackValue(lhs.album) == normalizedPlaybackValue(rhs.album)
      && abs(lhs.duration - rhs.duration) <= Self.staleProgressToleranceSeconds
  }

  private func normalizedPlaybackValue(_ value: String?) -> String {
    value?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
  }

  private func userFacingStatus(code: String?, message: String?) -> String {
    switch code {
    case "host_disabled":
      return "Connected Apps disabled"
    case "connector_blocked":
      return "Connector disabled"
    case "connector_not_registered":
      return "Reconnecting..."
    case "unsupported_protocol":
      return "Update required"
    default:
      break
    }

    if isMissingYouTubeMusicTabError(message) {
      return "No YouTube Music tab"
    }

    return message ?? code ?? "Connector error"
  }

  private func isMissingYouTubeMusicTabError(_ message: String?) -> Bool {
    message?.contains("Receiving end does not exist") == true
      || message?.contains("No active YouTube Music tab") == true
      || message?.contains("No YouTube Music tab") == true
      || message?.contains("No YTM tab") == true
  }

  private func logValue(_ value: String?) -> String {
    guard let value, !value.isEmpty else { return "nil" }
    return "\"\(value)\""
  }
}

File: apps/windows-tray/src/YTMTray.Core/ConnectorApp.cs
using System.Diagnostics;

namespace YTMTray.Core;

public sealed class ConnectorApp : IDisposable
{
    private static readonly TimeSpan DefaultPlaybackStateRetryDelay = TimeSpan.FromSeconds(2);
    private static readonly TimeSpan DefaultPlaybackStateStaleTimeout = TimeSpan.FromSeconds(8);
    private static readonly TimeSpan StaleProgressTolerance = TimeSpan.FromSeconds(0.25);

    private readonly IConnectorConnection connection;
    private readonly ITrayController tray;
    private readonly NativeAppLogger logger;
    private readonly Action openYouTubeMusic;
    private readonly TimeSpan playbackStateRetryDelay;
    private readonly TimeSpan playbackStateStaleDelay;
    private int nextRequestNumber;
    private bool ready;
    private bool disposed;
    private volatile bool playbackStateStale;
    private PlaybackState? lastAcceptedPlaybackState;
    private Timer? playbackStateRetry;
    private Timer? playbackStateStaleTimeout;

    public ConnectorApp(
        IConnectorConnection connection,
        ITrayController tray,
        NativeAppLogger? logger = null,
        TimeSpan? playbackStateRetryDelay = null,
        TimeSpan? playbackStateStaleTimeout = null,
        Action? openYouTubeMusic = null
    )
    {
        this.connection = connection;
        this.tray = tray;
        this.logger = logger ?? new NativeAppLogger();
        this.openYouTubeMusic =
            openYouTubeMusic ?? OpenYouTubeMusicInDefaultBrowser;
        this.playbackStateRetryDelay =
            playbackStateRetryDelay ?? DefaultPlaybackStateRetryDelay;
        playbackStateStaleDelay =
            playbackStateStaleTimeout ?? DefaultPlaybackStateStaleTimeout;
    }

    public void Start()
    {
        tray.OnShuffle = () => SendAction("shuffle");
        tray.OnPrevious = () => SendAction("previous");
        tray.OnTogglePlay = () => SendAction("togglePlay");
        tray.OnNext = () => SendAction("next");
        tray.OnRepeat = () => SendAction("repeat");
        tray.OnSeek = SendSeek;
        tray.OnFocusYouTubeMusic = SendFocusYouTubeMusic;

        logger.Log("connector app starting");
        connection.Start(Handle, HandleDisconnect);
        _ = connection.SendAsync(ConnectorProtocol.Hello(NextRequestId("hello")));
    }

    public void Dispose()
    {
        if (disposed) return;

        disposed = true;
        ClearPlaybackStateRetry();
        ClearPlaybackStateStaleTimeout();
        connection.Dispose();
    }

    private void Handle(HostMessage message)
    {
        logger.Log(
            $"handling message type={message.Type} requestId={message.RequestId ?? "none"}"
        );

        switch (message.Type)
        {
            case "connector.ready":
                ready = true;
                playbackStateStale = false;
                ClearPlaybackStateStaleTimeout();
                tray.UpdateConnectionStatus("Connected");
                tray.UpdateBrowserSource(message.Source);
                logger.Log(
                    $"connector ready source={message.Source?.DisplayName ?? "unknown"}"
                );
                _ = connection.SendAsync(
                    ConnectorProtocol.SubscribePlayback(NextRequestId("subscribe"))
                );
                RequestPlaybackState();
                RequestYtmStatus();
                break;
            case "playback.state":
                if (message.State is not null)
                {
                    HandlePlaybackState(message.State);
                }
                else
                {
                    logger.Log("playback state message missing state payload");
                }
                break;
            case "ytm.status":
                if (message.Status is not null)
                {
                    tray.UpdateYouTubeMusicTabAvailability(message.Status.HasTabs);
                }
                break;
            case "connector.error":
                HandleConnectorError(message);
                break;
            case ConnectorProtocol.UninstallRequestedType:
                logger.Log("connector uninstall requested");
                tray.RequestUninstall();
                break;
            default:
                logger.Log($"ignored message type={message.Type}");
                break;
        }
    }

    private void HandleDisconnect()
    {
        logger.Log("connector disconnected");
        ready = false;
        playbackStateStale = false;
        lastAcceptedPlaybackState = null;
        ClearPlaybackStateRetry();
        ClearPlaybackStateStaleTimeout();
        tray.UpdateConnectionStatus("Disconnected");
        tray.UpdateBrowserSource(null);
        tray.UpdateYouTubeMusicTabAvailability(false);

        if (!disposed)
        {
            _ = connection.SendAsync(ConnectorProtocol.Hello(NextRequestId("hello")));
        }
    }

    private void HandlePlaybackState(PlaybackState state)
    {
        ClearPlaybackStateRetry();
        ClearPlaybackStateStaleTimeout();
        tray.UpdateYouTubeMusicTabAvailability(true);
        logger.Log(PlaybackStateSummary(state));

        if (ShouldKeepStalePlaybackState(state))
        {
            logger.Log("playback state still stale; ignoring non-advancing poll");
            tray.SetStalePlaybackState();
            SchedulePlaybackStateStaleTimeout();
            return;
        }

        lastAcceptedPlaybackState = state;
        playbackStateStale = false;
        tray.UpdatePlayback(state);

        if (state.IsPlaying)
        {
            SchedulePlaybackStateStaleTimeout();
        }
    }

    private void HandleConnectorError(HostMessage message)
    {
        var label = UserFacingStatus(message.Code, message.Message);
        logger.Log($"connector error {label}");

        if (IsUnsupportedYtmStatusRequestError(message))
        {
            logger.Log("ignoring unsupported YouTube Music status request");
            return;
        }

        if (message.Code == "connector_not_registered")
        {
            RestartHandshake(label);
            return;
        }

        if (IsFocusYouTubeMusicRequestError(message))
        {
            tray.UpdateYouTubeMusicTabAvailability(false);
            OpenYouTubeMusicLocally("extension could not find a YouTube Music tab");
            return;
        }

        if (IsPlaybackStateRequestError(message))
        {
            tray.UpdateYouTubeMusicTabAvailability(false);
            tray.UpdateConnectionStatus(label);
            SchedulePlaybackStateRetry();
            return;
        }

        if (IsConnectorAvailabilityError(message.Code))
        {
            ready = false;
            playbackStateStale = false;
            lastAcceptedPlaybackState = null;
            ClearPlaybackStateRetry();
            ClearPlaybackStateStaleTimeout();
            tray.UpdateBrowserSource(null);
            tray.UpdateYouTubeMusicTabAvailability(false);
        }

        tray.UpdateConnectionStatus(label);
    }

    private void RequestPlaybackState()
    {
        if (!ready)
        {
            logger.Log("playback state refresh skipped; connector is not ready");
            return;
        }

        ClearPlaybackStateRetry();
        _ = connection.SendAsync(
            ConnectorProtocol.PlaybackStateRequest(NextRequestId("state"))
        );
    }

    private void RequestYtmStatus()
    {
        if (!ready) return;
        _ = connection.SendAsync(
            ConnectorProtocol.YtmStatusRequest(NextRequestId("ytm-status"))
        );
    }

    private void SendAction(string action)
    {
        if (!ready) return;
        _ = connection.SendAsync(
            ConnectorProtocol.PlaybackAction(action, NextRequestId("action"))
        );
    }

    private void SendSeek(double time)
    {
        if (!ready) return;
        _ = connection.SendAsync(ConnectorProtocol.PlaybackSeek(time, NextRequestId("seek")));
    }

    private void SendFocusYouTubeMusic()
    {
        if (!ready)
        {
            OpenYouTubeMusicLocally("connector is not ready");
            return;
        }
        _ = connection.SendAsync(ConnectorProtocol.FocusYouTubeMusic(NextRequestId("focus")));
    }

    private string NextRequestId(string prefix) =>
        $"{prefix}-{Interlocked.Increment(ref nextRequestNumber)}";

    private bool IsPlaybackStateRequestError(HostMessage message) =>
        message.RequestId?.StartsWith("state-", StringComparison.Ordinal) == true
        && IsMissingYouTubeMusicTabMessage(message.Message);

    private bool IsFocusYouTubeMusicRequestError(HostMessage message) =>
        message.RequestId?.StartsWith("focus-", StringComparison.Ordinal) == true
        && IsMissingYouTubeMusicTabMessage(message.Message);

    private static bool IsUnsupportedYtmStatusRequestError(HostMessage message) =>
        message.RequestId?.StartsWith("ytm-status-", StringComparison.Ordinal) == true
        && message.Code == "invalid_message";

    private void SchedulePlaybackStateRetry()
    {
        ClearPlaybackStateRetry();
        playbackStateRetry = new Timer(
            _ => RequestPlaybackState(),
            null,
            playbackStateRetryDelay,
            Timeout.InfiniteTimeSpan
        );
    }

    private void SchedulePlaybackStateStaleTimeout()
    {
        ClearPlaybackStateStaleTimeout();
        playbackStateStaleTimeout = new Timer(
            _ =>
            {
                playbackStateStale = true;
                tray.SetStalePlaybackState();
                RequestPlaybackState();
            },
            null,
            playbackStateStaleDelay,
            Timeout.InfiniteTimeSpan
        );
    }

    private void RestartHandshake(string reason)
    {
        ready = false;
        playbackStateStale = false;
        lastAcceptedPlaybackState = null;
        ClearPlaybackStateRetry();
        ClearPlaybackStateStaleTimeout();
        tray.UpdateConnectionStatus(reason);
        tray.UpdateBrowserSource(null);
        tray.UpdateYouTubeMusicTabAvailability(false);
        _ = connection.SendAsync(ConnectorProtocol.Hello(NextRequestId("hello")));
    }

    private void OpenYouTubeMusicLocally(string reason)
    {
        logger.Log($"opening YouTube Music in the default browser because {reason}");
        try
        {
            openYouTubeMusic();
        }
        catch (Exception error)
        {
            logger.Log($"could not open YouTube Music: {error.Message}");
            tray.UpdateConnectionStatus("Could not open YouTube Music");
        }
    }

    private static void OpenYouTubeMusicInDefaultBrowser()
    {
        Process.Start(
            new ProcessStartInfo("https://music.youtube.com/")
            {
                UseShellExecute = true
            }
        );
    }

    private bool ShouldKeepStalePlaybackState(PlaybackState state)
    {
        if (!playbackStateStale) return false;
        if (lastAcceptedPlaybackState is null) return false;
        if (state.Duration <= 0) return false;
        if (!lastAcceptedPlaybackState.IsPlaying || !state.IsPlaying) return false;
        if (!SamePlaybackItem(lastAcceptedPlaybackState, state)) return false;
        return Math.Abs(state.Progress - lastAcceptedPlaybackState.Progress)
            <= StaleProgressTolerance.TotalSeconds;
    }

    private static bool SamePlaybackItem(PlaybackState a, PlaybackState b) =>
        a.Title == b.Title
        && a.Artist == b.Artist
        && a.Album == b.Album
        && a.Year == b.Year
        && a.ArtworkUrl == b.ArtworkUrl
        && a.IsShuffling == b.IsShuffling
        && a.RepeatMode == b.RepeatMode
        && Math.Abs(a.Duration - b.Duration) <= StaleProgressTolerance.TotalSeconds
        && SameTrackMetadata(a.NextTrack, b.NextTrack);

    private static bool SameTrackMetadata(TrackMetadata? a, TrackMetadata? b)
    {
        if (a is null || b is null) return a is null && b is null;

        return a.Title == b.Title
            && a.Artist == b.Artist
            && a.Album == b.Album
            && a.Year == b.Year
            && a.ArtworkUrl == b.ArtworkUrl;
    }

    private static string PlaybackStateSummary(PlaybackState state) =>
        string.Join(
            " ",
            [
                "playback state",
                $"title={LogValue(state.Title)}",
                $"artist={LogValue(state.Artist)}",
                $"album={LogValue(state.Album)}",
                $"artwork={LogValue(state.ArtworkUrl)}",
                $"nextTrack={LogValue(state.NextTrack?.Title)}",
                $"nextTrackArtwork={LogValue(state.NextTrack?.ArtworkUrl)}",
                $"year={state.Year?.ToString() ?? "nil"}",
                $"playing={state.IsPlaying}",
                $"progress={Math.Round(state.Progress)}",
                $"duration={Math.Round(state.Duration)}",
                $"shuffle={state.IsShuffling?.ToString() ?? "nil"}",
                $"repeat={state.RepeatMode ?? "nil"}"
            ]
        );

    private static string LogValue(string? value) =>
        string.IsNullOrWhiteSpace(value) ? "nil" : value;

    private static bool IsConnectorAvailabilityError(string? code) =>
        code is "host_disabled" or "connector_blocked" or "unsupported_protocol";

    private static bool IsMissingYouTubeMusicTabMessage(string? message) =>
        message?.Contains("Receiving end does not exist", StringComparison.OrdinalIgnoreCase)
            == true
        || message?.Contains(
                "No active YouTube Music tab",
                StringComparison.OrdinalIgnoreCase
            )
            == true
        || message?.Contains("No YouTube Music tab", StringComparison.OrdinalIgnoreCase)
            == true
        || message?.Contains("No YTM tab", StringComparison.OrdinalIgnoreCase) == true;

    private static string UserFacingStatus(string? code, string? message) =>
        code switch
        {
            "host_disabled" => "Connected Apps disabled",
            "connector_blocked" => "Connector disabled",
            "unsupported_protocol" => "Update required",
            "connector_not_registered" => "Reconnecting...",
            _ when IsMissingYouTubeMusicTabMessage(message) => "No YouTube Music tab",
            _ => string.IsNullOrWhiteSpace(message) ? "Unavailable" : message
        };

    private void ClearPlaybackStateRetry()
    {
        playbackStateRetry?.Dispose();
        playbackStateRetry = null;
    }

    private void ClearPlaybackStateStaleTimeout()
    {
        playbackStateStaleTimeout?.Dispose();
        playbackStateStaleTimeout = null;
    }
}
